donkeycar/parts/gps.py

```diff
@@ -2,163 +2,190 @@
+                    positions.append((ts, position[0], position[1]))
+        return positions
 
-    def clear(self):
-        """
-        Clear the positions buffer
-        """
-        with self.lock:
-            try:
-                if self.gps is not None and self.gps.is_open:
-                    self.positions = []
-                    self.gps.reset_input_buffer()
-            except serial.serialutil.SerialException:
-                pass
+    def update(self):
+        pass
 
-    def _readline(self) -> str:
-        """
-        Read a line from the gps in a threadsafe manner
-        returns line if read and None if no line was read
-        """
-        if self.lock.acquire(blocking=False):
-            try:
-                # TODO: Serial.in_waiting _always_ returns 0 in Macintosh
-                if self.gps is not None and self.gps.is_open and (is_mac() or self.gps.in_waiting):
-                    return self.gps.readline().decode()
-            except serial.serialutil.SerialException:
-                pass
-            except UnicodeDecodeError:
-                # the first sentence often includes mis-framed garbase
-                pass  # ignore and keep going
-            finally:
-                self.lock.release()
-        return None
+    def run_threaded(self, lines):
+        return self.run(lines)
 
-    def poll(self, timestamp=None):
-        #
-        # read a line and convert to a position
-        # in a threadsafe manner
-        #
-        # if there are characters waiting
-        # then read the line and parse it
-        #
-        if self.running:
-            if timestamp is None:
-                timestamp = time.time()
-            line = self._readline()
-            if line:
-                if self.debug:
-                    logger.info(line)
-                position = getGpsPosition(line, debug=self.debug)
-                if position:
-                    # (timestamp, longitude latitude)
-                    return (timestamp, position[0], position[1])
-        return None
+class GpsLatestPosition:
+    """
+    Return most recent valid GPS position
+    """
+    def __init__(self, debug=False):
+        self.debug = debug
+        self.position = None
 
-    def run(self):
-        if self.running:
-            #
-            # in non-threaded mode, just read a single reading and return it
-            #
-            if self.gps is not None:
-                position = self.poll(time.time())
-                if position:
-                    # [(timestamp, longitude, latitude)]
-                    return [position]
+    def run(self, positions):
+        if positions is not None and len(positions) > 0:
+            self.position = positions[-1]
+        return self.position
 
-        return []
+class GpsPosition:
+    """
+    Donkeycar part to read NMEA lines from serial port and convert a position
+    """
+    def __init__(self, serial:SerialPort, debug = False) -> None:
```

Comment:
It would be better to use the logger's options to decide if logging is switched on or not.

Reply:
I don't want the rest of the debug logging when this is on.

Reply:
Do you know you can configure the logger to switch on different log levels for different modules? So you can set the loglevel at `donkeycar` to INFO which will trickle down to all modules but then you can set it in `donkeycar.parts.gps` to DEBUG?